external/fv3fit/fv3fit/emulation/thermobasis/loss.py

```diff
@@ -0,0 +1,169 @@
+import abc
+import dataclasses
+from typing import List, Mapping, Tuple, Any
+import tensorflow as tf
+from fv3fit.emulation.thermobasis.thermo import ThermoBasis
+from fv3fit.emulation.thermo import relative_humidity, specific_humidity_from_rh
+
+
+Info = Mapping[str, float]
+
+
+def rh_loss_info(truth_rh, pred_rh, level):
+    loss_rh = tf.reduce_mean(tf.losses.mean_squared_error(truth_rh, pred_rh))
+    return {
+        f"relative_humidity_mse/level_{level}": loss_rh.numpy() * (86400 / 900) ** 2
+    }
+
+
+def q_loss_info(truth_q, pred_q, level):
+    loss_q = tf.reduce_mean(tf.losses.mean_squared_error(truth_q, pred_q))
+    return {
+        f"loss/variable_3/level_{level}": loss_q.numpy() * (1000 * 86400 / 900) ** 2
+    }
+
+
+class Loss(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def loss(self, prediction: Any, out: ThermoBasis) -> Tuple[tf.Tensor, Info]:
+        """Compute a loss value and corresponding information
+
+        Returns:
+            loss, info: loss is a backwards differntiable scalar value used for
+                gradient descent, info is a dictionary of details about the loss
+                value calculation (e.g. the MSEs of relative humidity and specific
+                humidity).
+        """
+        pass
+
+
+@dataclasses.dataclass
+class QVLossSingleLevel:
+    """Loss function for predicting specific humidity **at a single level**
+
+    Attributes:
+        level: the level to predict
+        scale: the typical order of the loss function
+    """
+
+    level: int
+    scale: float = 1.0
+
+    def loss(self, pred: tf.Tensor, out: ThermoBasis) -> Tuple[tf.Tensor, Info]:
+        truth_q = select_level(out.q, self.level)
+        loss = tf.reduce_mean(tf.losses.mean_squared_error(truth_q, pred))
+
+        pred_rh = relative_humidity(
+            select_level(out.T, self.level), pred, select_level(out.rho, self.level),
+        )
+        truth_rh = select_level(out.rh, self.level)
+
+        return (
+            loss / self.scale,
+            {
+                **q_loss_info(truth_q, pred, self.level),
+                **rh_loss_info(truth_rh, pred_rh, self.level),
+            },
+        )
+
+
+@dataclasses.dataclass
+class RHLossSingleLevel:
+    """Loss function for predicting relative humidity **at a single level**
+
+    Attributes:
+        variable: the variable to target, defaults to all levels of u,v,t,q
+        level: the level to predict
+        scale: the typical order of the loss function
+    """
+
+    level: int
+    scale: float = 1.0
+
+    def loss(self, pred_rh: tf.Tensor, out: ThermoBasis) -> Tuple[Loss, Info]:
```

Comment:
Thanks for the class docstring! Can you please add a docstring to the loss function specifying whether these should have data from a single level or all levels?
```suggestion
    def loss(self, pred_rh: tf.Tensor, out: ThermoBasis) -> Tuple[Loss, Info]:
        """
        Args:
            pred_rh: predicted relative humidity at a single vertical level
            out: target values at all vertical levels
        """
```

Reply:
See bec9b83.